independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanArchives.java

```diff
@@ -0,0 +1,240 @@
+package io.quarkus.arc.processor;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.enterprise.context.BeforeDestroyed;
+import javax.enterprise.context.Destroyed;
+import javax.enterprise.context.Initialized;
+import javax.enterprise.context.control.ActivateRequestContext;
+import javax.enterprise.inject.Any;
+import javax.enterprise.inject.Default;
+import javax.inject.Named;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.CompositeIndex;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.Index;
+import org.jboss.jandex.IndexView;
+import org.jboss.jandex.Indexer;
+import org.jboss.jandex.Type;
+
+import io.quarkus.arc.ActivateRequestContextInterceptor;
+
+public final class BeanArchives {
+
+    /**
+     * 
+     * @param applicationIndexes
+     * @return the final bean archive index
+     */
+    public static IndexView buildBeanArchiveIndex(IndexView... applicationIndexes) {
+        List<IndexView> indexes = new ArrayList<>();
+        Collections.addAll(indexes, applicationIndexes);
+        // Add CDI API builtin annotations if needed
+        if (!contains(DotNames.ANY, applicationIndexes)) {
+            indexes.add(buildCdiApiIndex());
+        }
+        if (!contains(DotName.createSimple(ActivateRequestContextInterceptor.class.getName()), applicationIndexes)) {
+            indexes.add(buildBuiltinIndex());
+        }
+        return new IndexWrapper(CompositeIndex.create(indexes));
+    }
+
+    private static boolean contains(DotName className, IndexView... indexes) {
+        for (IndexView index : indexes) {
+            if (index.getClassByName(className) != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static IndexView buildCdiApiIndex() {
+        Indexer indexer = new Indexer();
+        index(indexer, ActivateRequestContext.class.getName());
+        index(indexer, Default.class.getName());
+        index(indexer, Any.class.getName());
+        index(indexer, Named.class.getName());
+        index(indexer, Initialized.class.getName());
+        index(indexer, BeforeDestroyed.class.getName());
+        index(indexer, Destroyed.class.getName());
+        return indexer.complete();
+    }
+
+    private static IndexView buildBuiltinIndex() {
+        Indexer indexer = new Indexer();
+        index(indexer, ActivateRequestContextInterceptor.class.getName());
+        return indexer.complete();
+    }
+
+    /**
+     * This wrapper is used to index JDK classes on demand.
+     */
+    static class IndexWrapper implements IndexView {
```

Comment:
This is new, isn't it?

Is it exposed to the extensions (i.e. is this wrapper part of the `IndexView` we get in the extensions)?

Reply:
It's not new. Previously it was a part of the `BeanDeployment` class. And the original purpose was to make JDK classes part of the index whenever ArC asks for a `ClassInfo` of a `DotName` that is not present in the provided index (for example a producer method returns a JDK `List` which is obviously not part of the app index but we need to analyze the corresponding `ClassInfo`). Of course, it works for non-JDK classes too.

Also it's package-private so it's not exposed. Anyway, extension can access this wrapper through the `IndexView` interface (after injecting `BeanArchiveIndexBuildItem`).